docs/rules/function-component-definition.md

```diff
@@ -0,0 +1,228 @@
+
+```js
+...
+"react/function-component-definition": [<enabled>, { 
+  "namedComponents": "function-declaration"|"function-expression"|"arrow-function",
+  "unnamedComponents": "function-expression"|"arrow-function"
+}]
+...
+```
+
+The following patterns are considered warnings:
+
+```jsx
+// only function declarations for named components
+// [2, { "namedComponents": "function-declaration" }]
+const Component = function (props) {
+  return <div/>
+}
+
+const Component = (props) => {
+  return <div/>
+}
+
+// only function expressions for named components
+// [2, { "namedComponents": "function-expression" }]
+function Component (props) {
+  return <div/>
+}
+
+const Component = (props) => {
+  return <div/>
+}
+
+// only arrow functions for named components
+// [2, { "namedComponents": "arrow-function" }]
+function Component (props) {
+  return <div/>
+}
+
+const Component = function (props) {
+  return <div/>
+}
+
+// only function expressions for unnamed components
+// [2, { "unnamedComponents": "function-expression" }]
+function getComponent () {
+  return (props) => {
+    return <div/>
+  }
+}
+
+// only arrow functions for unnamed components
+// [2, { "unnamedComponents": "arrow-function" }]
+function getComponent () {
+  return (props) => {
+    return <div/>
+  }
+}
+
+```
+
+The following patterns are **not** warnings:
+
+```jsx
+
+// only function declarations for named components
+// [2, { "namedComponents": "function-declaration" }]
+function Component (props) {
+  return <div/>
+}
+
+// only function expressions for named components
+// [2, { "namedComponents": "function-expression" }]
+const Component = function (props) {
+  return <div/>
+}
+
+// only arrow functions for named components
+// [2, { "namedComponents": "arrow-function" }]
+const Component = (props) => {
+  return <div/>
+}
+
+// only function expressions for unnamed components
+// [2, { "unnamedComponents": "function-expression" }]
+function getComponent () {
+  return (props) => {
+    return <div/>
+  }
+}
+
+// only arrow functions for unnamed components
+// [2, { "unnamedComponents": "arrow-function" }]
+function getComponent () {
+  return (props) => {
+    return <div/>
+  }
+}
+
+```
+
+## Heads up typescript users
+Note that the auto fixer is somewhat constrained for typescript users. 
+
+The following patterns can **not** be autofixed in typescript:
+
+```tsx
+// function expressions and arrow functions that have type annotations cannot be autofixed to function declarations
+// [2, { "namedComponents": "function-declaration" }]
+const Component: React.FC<Props> = function (props) {
+  return <div/>
+}
+
+const Component: React.FC<Props> = (props) => {
+  return <div/>
+}
+
+// function components with one unconstrained type parameter cannot be autofixed to arrow functions because the syntax conflicts with jsx
+// [2, { "namedComponents": "arrow-function" }]
+function Component<T>(props: Props<T>) {
```

Comment:
why can't this be autofixed? there can't be jsx in these positions.

Reply:
It's probably a bug in typescript but in tsx files this cannot be parsed correctly.